src/coreclr/vm/unsafeaccessors.cpp

```diff
@@ -477,6 +704,59 @@ namespace
         return false;
     }
 
+    void EmitTypeCheck(ILCodeStream* pCode, TypeHandle th)
+    {
+        STANDARD_VM_CONTRACT;
+
+        if (th.IsNull())
+            return;
+
+        TypeHandle origType = th;
+
+        // We are going to emit a type check for the UnsafeAccessorTypeAttribute
+        // scenario. We do this because the callable signature isn't going to enforce
+        // type safety, so we do it here.
+        //
+        // If the type is a byref, DUP the value so the execution stack has a
+        // byref value on it after the type check. Emit a load indirection of the byref
+        // in order to validate the type. After the typecheck, we'll pop off the object.
+        // This doesn't handle the null byref case.
```

Comment:
This has type safety hole. It does not handle the case where the value that the byref is pointing to is changed between this check and use of the byref in the target method.

Reply:
Can you show me an example of how this can occur? I'm not sure I understand how a byref can change in this manner. I'm interpreting this statement as follows.

```csharp

ref obj = ref ...

M(new A(), ref obj); // Is a another thread changing the object that at the ref?

extern static void M(A a, [UnsafeAccessorType("B&")] ref object o);
```

I considered the above case but there is a trade-off and I took the case that changing the value of the actual `ref` could be detected. So what I had done was create a fully typed local in the stub and pass a `ref` to that and then update it. I'm not sure which is worse. Both are detectable but the thread one seems far less common than someone expecting the `ref` to be the same.

Reply:
You can hit the type safety hole on a single thread as well. Something like:
```
using System.Runtime.CompilerServices;

class Test
{
    static object o;

    static void Main()
    {
        o = "Hello";
        M(ref Unsafe.As<object, string>(ref o)); // Unsafe.As is used to simulate what's done by the UnsafeAccessor
    }

    static void M(ref string s)
    {
       o = new Test();
       Console.WriteLine(s.GetType());
    }
}
```
One option is to omit support for `ref`s for now and leave it for the future. `TypedReference`s allow passing of refs in a type-safe way.

Reply:
Bah. The global state here seems a bit contrived, but I understand the example. It does seem more likely than the threading one.

> One option is to omit support for refs for now and leave it for future. TypedReferences allow passing of refs in a type-safe way.

I can do that or I can do the local approach I originally had. I will create a typed local and pass a ref to that and then update the ref on the out. Thoughts?

Reply:
Is anybody asking for it right now? Leave it unsupported for now is the easy option.

Reply:
I can't think of a scenario where we would need `[UnsafeAccessorType]` on a byref parameter in CsWinRT off the top of my head, no. I can see how genetics might potentially be useful in some scenarios (eg. passing a derived generic type as parameter to some static method in the sidecar, or something), but can't imagine where we'd need any of those params to be byrefs. So basically dropping this (byrefs params) should be completely fine from the CsWinRT side 🙂

Reply:
I should have stated this question more accurately. From the runtime perspective byrefs are expressed as `in`, `ref` and `out` in C#. None of these would be permitted without this support.

> TypedReferences allow passing of refs in a type-safe way.

@jkotas I thought a bit more about this. I don't see how a `TypedReference` is going to address the motivating example. The API is still going to accept a `ref string` and the `static` will still be `object` and assigned a `Test` instance. The `TypedReference` isn't going to stop any of those actions and the `ref` contained within the `TypedReference` instance will now point at a `Test`, but typed as a `string`. What am I missing here?

Reply:
I assume that the user signature would need to take a `TypedReference` argument.

The caller wouldn't be able to pass a ref to `o` as a string-typed `TypedReference`. They'd have to construct a `TypedReference` to a `string`-typed stack location.

If the user were to construct a string-typed `TypedReference` to the location of `o`, that would require the user to write type-unsafe code (ie. using `Unsafe.As` in their code to get a `ref string` from a `ref object`).

Reply:
Thanks @jkoritzinsky. Yes, that is what was intended here and I didn't understand that assumption. We discuss offline and I understand the assumption. I'm going to follow the reflection approach and make a local in the stub and pass a reference to that and then also on the out side.

Reply:
Okay. I'll do something similar for the NAOT implementation.